Allow using structs with trivial initializers in globals that require static initialization (e.g. @_section attribute)

kubamracek opened the pull request:
Currently, if a global variable is required to be statically initialized (e.g. due to @_section attribute), we don't allow its type to be a structs, only a scalar type works. Let's improve on that by teaching MandatoryPerformanceOptimizations pass to inline struct initializer calls into initializer of globals, as long as they are simple enough so that we can be sure that we don't trigger recursive/infinite inlining.

eeckstein commented on SwiftCompilerSources/Sources/SIL/GlobalVariable.swift:
```diff
@@ -53,6 +53,10 @@ final public class GlobalVariable : CustomStringConvertible, HasShortDescription
     return bridged.canBeInitializedStatically()
   }
 
+  public var mustBeInitializedStatically: Bool {
```
I would add a `hasSectionAttribute` property (which bridges to C++) and check this property here. Then it's immediately clear that a global must be initialized statically if it has an section attribute.
Also, we can add other stuff in the future which might require static initialization

kubamracek replied:
I considered that, but then realized that a mustBeInitializedStatically() method is useful from existing C++ code as well. If I implement it on the Swift side, will I be able to call it from e.g. PerformanceDiagnostics.cpp?

eeckstein commented on test/IRGen/section_structs.swift:
```diff
@@ -0,0 +1,45 @@
+// RUN: %target-swift-frontend -enable-experimental-feature SymbolLinkageMarkers -parse-as-library -emit-ir %s -o - | %FileCheck %s
+
+// REQUIRES: swift_in_compiler
+
```
Please also add generic structs, because the SIL for initializers for generic structs looks much different than for non generic structs.

kubamracek replied:
Added in section_structs_generic.swift

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/ModulePasses/MandatoryPerformanceOptimizations.swift:
```diff
@@ -146,6 +181,20 @@ fileprivate struct FunctionWorklist {
 
+  mutating func addAllAnnotatedGlobalInitOnceFunctions(of moduleContext: ModulePassContext) {
+    for f in moduleContext.functions where f.isGlobalInitOnceFunction {
+      guard let global = getGlobal(of: f) else {
```
nitpick: Instead of multiple guard statements you can make a single if statement, combining all the conditions, e.g.
```
  if let global = f. getInitializedGlobal(),
     global.mustBeInitializedStatically {
    pushIfNotVisited(f)
  }
```
IMO, this reads more naturally.

eeckstein reviewed (commented):
nice! I have a few comments

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/ModulePasses/MandatoryPerformanceOptimizations.swift:
```diff
@@ -123,9 +123,44 @@ private func shouldInline(apply: FullApplySite, callee: Function) -> Bool {
   return false
 }
 
+private func resultOfCalleeAssignsAMember(apply: FullApplySite) -> Bool {
```
If we want to support things like
```
init() {
  let x = X()
  self.a = x
  self.b = x
}
```
Then a loop is not sufficient. It needs to be a recursive function, visiting _all_ uses rather than the singleUse.

kubamracek commented on SwiftCompilerSources/Sources/Optimizer/ModulePasses/MandatoryPerformanceOptimizations.swift:
```diff
@@ -123,9 +132,66 @@ private func shouldInline(apply: FullApplySite, callee: Function) -> Bool {
+          return false
+        }
+        
+        return /*goesThroughAMemberInit &&*/ storeDestination.global == global
```
The "require that the def-use chain goes through a struct/tuple/enum inst" (goesThroughAMemberInit) part unfortunately seems to reject some simple situations where we would want the inlining to happen, e.g. this SIL:
```
// one-time initialization function for g_MyStruct2
sil private [global_init_once_fn] [ossa] @$s4null11g_MyStruct2_WZ : $@convention(c) (Builtin.RawPointer) -> () {
bb0(%0 : $Builtin.RawPointer):
  alloc_global @$s4null11g_MyStruct2AA0bC0Vvp     // id: %1
  %2 = global_addr @$s4null11g_MyStruct2AA0bC0Vvp : $*MyStruct2 // user: %8
  %3 = metatype $@thin MyStruct2.Type             // user: %7
  %4 = integer_literal $Builtin.Int64, 42         // user: %5
  %5 = struct $Int (%4 : $Builtin.Int64)          // user: %7
  // function_ref MyStruct2.init(a:)
  %6 = function_ref @$s4null9MyStruct2V1aACSi_tcfC : $@convention(method) (Int, @thin MyStruct2.Type) -> MyStruct2 // user: %7
  %7 = apply %6(%5, %3) : $@convention(method) (Int, @thin MyStruct2.Type) -> MyStruct2 // user: %8
  store %7 to [trivial] %2 : $*MyStruct2          // id: %8
  %9 = tuple ()                                   // user: %10
  return %9 : $()                                 // id: %10
} // end sil function '$s4null11g_MyStruct2_WZ'
```
Ideas?

eeckstein replied:
You are right.
Let's drop that requirement for now. I have another idea to prevent infinite inlining, but that's a bit more complicate.

kubamracek replied:
PR is redone to use the SmallProjectionPath-based tracking of already inlined functions.

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/ModulePasses/MandatoryPerformanceOptimizations.swift:
```diff
@@ -48,28 +50,32 @@ private func optimizeFunctionsTopDown(using worklist: inout FunctionWorklist,
+  while changed {
+    changed = false
+    
+    changed = changed || runSimplification(on: function, context, preserveDebugInfo: true) { instruction, simplifyCtxt in
```
Is it really required to retrigger another iteration for all the things we do here?
I think it's only required if `optimizeMemoryAccesses` does some changes. Only this optimization can open more simplification opportunities.

kubamracek replied:
It seems like I need to at least check if runSimplification did anything *or* optimizeMemoryAccesses did anything, it's not enough to just check optimizeMemoryAccesses. Does that sound about right? I can find out why exactly.

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/ModulePasses/MandatoryPerformanceOptimizations.swift:
```diff
@@ -48,28 +50,32 @@ private func optimizeFunctionsTopDown(using worklist: inout FunctionWorklist,
-  if context.optimizeMemoryAccesses(in: function) {
-    _ = context.eliminateDeadAllocations(in: function)
+    
+    changed = changed || context.specializeApplies(in: function, isMandatory: true)
```
The `changed = changed || ...` does not work, because if `changed` is already true then the right hand side is not evaluated

kubamracek replied:
Ah, of course...

kubamracek committed:
Allow using structs with trivial initializers in globals that require static initialization (e.g. @_section attribute)

Before this change, if a global variable is required to be statically initialized (e.g. due to @_section attribute), we don't allow its type to be a struct, only a scalar type works. This change improves on that by teaching MandatoryPerformanceOptimizations pass to inline struct initializer calls into initializer of globals, as long as they are simple enough so that we can be sure that we don't trigger recursive/infinite inlining.

kubamracek commented:
@swift-ci please test

kubamracek committed:
Add section_structs_generic.swift to check constant-folding of generic struct in global variables

kubamracek commented:
@swift-ci please test

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/ModulePasses/MandatoryPerformanceOptimizations.swift:
```diff
@@ -123,9 +129,104 @@ private func shouldInline(apply: FullApplySite, callee: Function) -> Bool {
+        path = path.push(.tupleField, index: use.index)
+        break
+      case is EnumInst:
+        path = path.push(.enumCase, index: use.index)
```
`use.index` is wrong here. The `enum` instruction always has a single operand (in case of a payload). You need to push `EnumInst.caseIndex`.

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/ModulePasses/MandatoryPerformanceOptimizations.swift:
```diff
@@ -123,9 +129,104 @@ private func shouldInline(apply: FullApplySite, callee: Function) -> Bool {
+        
+        return path
+      default:
+        break
```
Is it okay to accept any kind of single-value instruction? Or should we be on the safe side and only accept `struct`, `tuple` and `enum`?

kubamracek replied:
Hm, sounds good -- turns out all the testcases I added still pass even if we only allow struct/tuple/enum. I guess I was trying to support things some simple arithmetic expressions inside the initializers, but I don't really see the immediate need for it.

kubamracek committed:
MandatoryPerformanceOptimizations: Use Set<PathFunctionTuple> to track inlined functions

kubamracek commented on SwiftCompilerSources/Sources/Optimizer/ModulePasses/MandatoryPerformanceOptimizations.swift:
```diff
@@ -48,33 +50,37 @@ private func optimizeFunctionsTopDown(using worklist: inout FunctionWorklist,
-      i.simplify(simplifyCtxt)
-      if instruction.isDeleted {
-        return
+  var alreadyInlinedFunctions: [SmallProjectionPath:Set<Function>] = [:]
```
Done, had to define a struct for it because the tuple isn't hashable automatically.

kubamracek commented:
@swift-ci please test

eeckstein reviewed (approved):
awesome! LGTM
